swim/state_transitions.go

```diff
@@ -0,0 +1,195 @@
+}
+
+// StateTimeouts contains the timeouts for every state before transitioning to the new state
+type StateTimeouts struct {
+	// Suspect is the timeout it takes a node in suspect mode to transition to faulty
+	Suspect time.Duration
+}
+
+func mergeStateTimeouts(one StateTimeouts, two StateTimeouts) StateTimeouts {
+	return StateTimeouts{
+		Suspect: util.SelectDuration(one.Suspect, two.Suspect),
+	}
+}
+
+// newStateTransitions returns a new state transition controller that can be used to schedule state transitions for nodes
+func newStateTransitions(n *Node, timeouts StateTimeouts) *stateTransitions {
+	return &stateTransitions{
+		node:     n,
+		timeouts: timeouts,
+		timers:   make(map[string]*transitionTimer),
+		enabled:  true,
+		logger:   logging.Logger("stateTransitions").WithField("local", n.Address()),
+	}
+}
+
+// ScheduleSuspectToFaulty starts the suspect timer. After the Suspect timeout the node will be declared faulty
+func (s *stateTransitions) ScheduleSuspectToFaulty(subject subject) {
+	s.Lock()
+	s.schedule(subject, Suspect, s.timeouts.Suspect, func() {
+		// transition the subject to faulty
+		s.node.memberlist.MakeFaulty(subject.address(), subject.incarnation())
+	})
+	s.Unlock()
+}
+
+func (s *stateTransitions) schedule(subject subject, state string, timeout time.Duration, transition func()) {
+	if !s.enabled {
+		s.logger.WithField("member", subject.address()).Warn("cannot schedule a state transition while disabled")
+		return
+	}
+
+	if s.node.Address() == subject.address() {
+		s.logger.WithField("member", subject.address()).Warn("cannot schedule a state transition for the local member")
+		return
+	}
+
+	if timer, ok := s.timers[subject.address()]; ok {
+		if timer.state == state {
+			s.logger.WithFields(log.Fields{
+				"member": subject.address(),
+				"state":  state,
+			}).Warn("redundant call to schedule a state transition for memer, ignored")
+			return
+		}
+		// cancel the previously scheduled transition for the subject
+		timer.Stop()
+	}
+
+	timer := time.AfterFunc(timeout, func() {
+		s.logger.WithFields(log.Fields{
+			"member": subject.address(),
+			"state":  state,
+		}).Info("executing scheduled transition for member")
+		// execute the transition
+		transition()
+	})
+
+	s.timers[subject.address()] = &transitionTimer{
+		Timer: timer,
+		state: state,
+	}
+
+	s.logger.WithFields(log.Fields{
+		"member": subject.address(),
+		"state":  state,
+	}).Debug("scheduled state transition for member")
+}
+
+// Cancel cancels the scheduled transition for the subject
+func (s *stateTransitions) Cancel(subject subject) {
+	s.Lock()
+
+	if timer, ok := s.timers[subject.address()]; ok {
+		timer.Stop()
+		delete(s.timers, subject.address())
+		s.logger.WithFields(log.Fields{
+			"member": subject.address(),
+			"state":  timer.state,
+		}).Debug("stopped scheduled state transition for member")
+	}
+
+	s.Unlock()
+}
+
+// Reenable state transition controller
+func (s *stateTransitions) Reenable() {
```

Comment:
`Disable` covers in its documentation that it cancels the outstanding timers.

I don't really know why they are in there, but they are called from `Node` on `Start` and `Stop`. It might have to do with enabling and disabling the gossip protocol which you can do from the admin endpoints.